Promote VolumeSubpathEnvExpansion e2e tests to Conformance

kevtaylor opened the pull request:
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
VolumeSubpathEnvExpansion is GA in 1.17 so the e2e tests can be added to Conformance suite

**Which issue(s) this PR fixes**:
Fixes #85122

**Special notes for your reviewer**:


**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

Please use the following format for linking documentation:
[KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20181029-volume-subpath-env-expansion.md
[Usage]: kubernetes/website#13846
[Other doc]: kubernetes/enhancements#559

kevtaylor committed:
Promote VolumeSubpathExpansion e2e tests to Conformance

kevtaylor commented:
/sig testing
/sig node
/sig storage

kevtaylor commented:
/cc @msau42

msau42 commented:
/ok-to-test
/assign

kevtaylor commented:
/retest

kevtaylor commented:
@msau42 I think that failing test is this https://github.com/kubernetes/kubernetes/issues/84940

msau42 commented:
I think there were issues in the past where we couldn't promote tests to conformance that needed to write to a hostpath volume. @spiffxp do you know if that's still the case?

@kevtaylor is it possible to use emptydir in these tests?

kevtaylor commented:
@msau42 I tested locally with EmptyDir which seems to work okay

kevtaylor commented:
/retest

msau42 commented on test/e2e/common/expansion.go:
```diff
@@ -158,7 +158,7 @@ var _ = framework.KubeDescribe("Variable Expansion", func() {
 			expansion of environment variables.
 	*/
-	ginkgo.It("should allow substituting values in a volume subpath [sig-storage]", func() {
+	framework.ConformanceIt("should allow substituting values in a volume subpath [sig-storage]", func() {
```
Thanks for updating! I think we'll need to do this as a two-step process. 1) Update to emptydir. 2) Once we verify it's stable and not flaky, then promote to conformance.

kevtaylor replied:
I have raised a separate Issue https://github.com/kubernetes/kubernetes/issues/87810 and will hold this one for the moment

kevtaylor replied:
@msau42 Haven't seen any errors in sig testgrid since the HostPath change, so I would guess we're good to go with this one now

kevtaylor replied:
I don't see anywhere obvious in the PR tests that has actually performed the conformance prior to merge. Is there any easy way of checking that this PR did some Conformance e2e?

kevtaylor replied:
I found that this one `pull-kubernetes-conformance-kind-ipv6-parallel` has performed all 6 conformance tests, so that seems to confirm they are okay

kevtaylor commented:
/hold

kevtaylor commented:
/unhold

kevtaylor commented:
/retest

kevtaylor commented:
/retest

kevtaylor commented:
/retest

msau42 commented:
/lgtm
/assign @johnbelamaric 
/retest

johnbelamaric commented:
Please take a look at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements and make sure these meet all the requirements before I dig in for a full review. One thing missing at first glance is the `Release:` meta-data.

Thanks.

johnbelamaric commented:
/priority important-soon

kevtaylor commented:
@johnbelamaric As far as I can see reading the documentation, these satisfy those constraints, particularly as we removed any hard dependencies on UNIX host in a previous PR. I also added the missing Release tag to the meta data

kevtaylor commented:
/retest

kevtaylor commented:
@johnbelamaric And the existing tests running in the grid are documented in the kep graduation criteria https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20181029-volume-subpath-env-expansion.md#graduation-criteria

kevtaylor commented:
There seems to be some new conformance.yaml file and the old text file has been deleted. Looking at this yaml file I cannot seem to see the exact match between the definitions and the code. There are differences. It also looks like the description fields are wrapped at some width boundary. Can anyone please shed some light on what I am to do with this?

kevtaylor commented:
@johnbelamaric I saw that there is a new regime for conformance.txt which has been deleted. According to the new conformance instructions README the yaml file is at 
`cp bazel-genfiles/test/conformance/conformance.yaml test/conformance/testdata`
but when I ran the bazel update it was in `bazel-bin`
```
Target //test/conformance:list_conformance_tests up-to-date:
  bazel-bin/test/conformance/conformance.yaml
```
Hopefully, that's the correct one

kevtaylor commented:
/retest

kevtaylor commented:
/retest

kevtaylor commented:
/retest

msau42 commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -410,6 +410,50 @@
+    subpath [sig-storage] [Conformance]'
+  description: Make sure a container's subpath can be set using an expansion of environment
+    variables.
+  release: v1.18
```
Update to 1.19

msau42 commented:
This lgtm, just have one comment to update the release since we missed the 1.19 window.
/approve

msau42 commented:
/lgtm
@johnbelamaric can you take another look?

kevtaylor commented:
/retest

kevtaylor commented:
/retest

kevtaylor commented:
/retest

kevtaylor commented:
/retest

johnbelamaric commented:
/approve
